build_projects/dotnet-cli-build/GitHubWriteVersionUpdater.cs

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#if !SOURCE_BUILD
+using Microsoft.DotNet.VersionTools.Automation;
+using Microsoft.DotNet.VersionTools.Automation.GitHubApi;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Net.Http;
+using System.Threading.Tasks;
+
+namespace Microsoft.DotNet.Cli.Build
+{
+    // This code is a slightly modified port of GitHubVersionsRepoUpdater from 
+    // https://raw.githubusercontent.com/dotnet/buildtools/master/src/Microsoft.DotNet.VersionTools/Automation/GitHubVersionsRepoUpdater.cs
+    internal class GitHubWriteVersionUpdater: VersionsRepoUpdater
```

Comment:
> How long have we been using this "client directly edit a repo strategy"?

Around two years.

> What happen if 2 build machines try to write different version at the same time aka merge conflict with build machines?

1. The machine that gets the `PostCommitAsync` call off first posts the the first commit.
2. The second machine detects that it can't push its change since the underlying commit has has changed, and retries.
   * The second machine's retry is based on the first machine's successful commit, so there is no git conflict.
3. The second machine calls `PostCommitAsync` and that change also gets pushed.

So there isn't really any conflict, just a forced ordering. If many builds were to complete at once, it would be essentially random based on which machine retries at the right time to grab the "lock". Practically, this has never been a problem.